src/source_file.rs

```diff
@@ -65,15 +74,25 @@ where
         }
     }
 
-    // If parse session is around (cfg(not(test))) then try getting source from
-    // there instead of hitting the file system. This also supports getting
+    // SourceFile's in the SourceMap will always have Unix-style line endings
+    // See: https://github.com/rust-lang/rustfmt/issues/3850
+    // So if the user has explicitly overridden the rustfmt `newline_style`
+    // config and `filename` is FileName::Real, then we must check the file system
+    // to get the original file value in order to detect newline_style conflicts.
+    // Otherwise, parse session is around (cfg(not(test))) and newline_style has been
+    // left as the default value, then try getting source from the parse session
+    // source map instead of hitting the file system. This also supports getting
     // original text for `FileName::Stdin`.
-    let original_text = source_map
-        .and_then(|x| x.get_source_file(&filename.into()))
-        .and_then(|x| x.src.as_ref().map(ToString::to_string));
-    let original_text = match original_text {
-        Some(ori) => ori,
-        None => fs::read_to_string(ensure_real_path(filename))?,
+    let original_text = if newline_style != NewlineStyle::Auto && *filename != FileName::Stdin {
+        fs::read_to_string(ensure_real_path(filename))?
+    } else {
+        match source_map
+            .and_then(|x| x.get_source_file(&filename.into()))
+            .and_then(|x| x.src.as_ref().map(ToString::to_string))
+        {
+            Some(ori) => ori,
+            None => fs::read_to_string(ensure_real_path(filename))?,
+        }
     };
```

Comment:
This change is to resolve the other bug I found while working on the diff emitter. It turns out that internally, as the parser constructs the source map, all of the files loaded into the source map store a value of the original file contents with all line endings normalized to Unix style:

https://github.com/rust-lang/rust/blob/31d75c4e9c5318e880601d3c2cc71e5df094a120/src/libsyntax_pos/lib.rs#L1065-L1072

https://github.com/rust-lang/rust/blob/53b352edb6f441bc3cf5386806fcc4686f275130/src/libsyntax/source_map.rs#L223-L269

This means that even if the original file contents were using Windows style line endings, that the Source Map will **always** represent the SourceFile.src values with Unix style line endings. As mentioned in the PR description, this results in buggy behavior with `rustfmt` whenever Windows style line endings are involved (either in rustfmt `newline_style` config and/or in the original file contents). In such instances, it is necessary to read the original file contents from disk in order to get a "correct" value for `original_text`

I'm of two minds as to whether this change should be version gated (and/or should require an explicit config option), but I ended up settling with this because my thinking was that users that are explicitly overriding the default `newline_style` config value in `rustfmt` would definitely expect the behavior this change provides. 

Definitely open to any suggested changes/alternatives though!